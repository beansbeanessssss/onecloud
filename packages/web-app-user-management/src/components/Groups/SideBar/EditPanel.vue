<template>
  <div class="oc-mt-xl">
    <div class="oc-flex group-info oc-mb-l">
      <avatar-image class="oc-mb-m" :width="80" :userid="group.id" :user-name="group.displayName" />
      <span class="oc-text-muted group-info-display-name" v-text="group.displayName"></span>
    </div>
    <div v-if="editGroup" class="oc-background-highlight oc-p-m">
      <oc-text-input
        v-model="editGroup.displayName"
        class="oc-mb-s"
        :label="$gettext('Group name')"
        :error-message="formData.displayName.errorMessage"
        :fix-message-line="true"
        @input="validateDisplayName"
      />
      <oc-button @click="$emit('confirm', editGroup)">ok</oc-button>
    </div>
  </div>
</template>
<script>
export default {
  name: 'EditPanel',
  props: {
    group: {
      type: Object,
      required: true
    }
  },
  data() {
    return {
      editGroup: {},
      formData: {
        displayName: {
          errorMessage: '',
          valid: true
        }
      }
    }
  },
  watch: {
    group: {
      handler: function () {
        this.editGroup = { ...this.group }
      },
      deep: true,
      immediate: true
    }
  },
  methods: {
    validateDisplayName() {
      this.formData.displayName.valid = false

      if (this.editGroup.displayName.trim() === '') {
        this.formData.displayName.errorMessage = this.$gettext('Display name cannot be empty')
        return false
      }

      this.formData.displayName.errorMessage = ''
      this.formData.displayName.valid = true
      return true
    }
  }
}
</script>
<style lang="scss">
.group-info {
  align-items: center;
  flex-direction: column;
}
.group-info-display-name {
  font-size: 1.5rem;
}
</style>
