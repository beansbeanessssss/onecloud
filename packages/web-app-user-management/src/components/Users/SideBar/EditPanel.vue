<template>
  <div class="oc-mt-xl">
    <UserInfoBox :user="user" />
    <div class="oc-background-highlight oc-p-m">
      <div>
        <oc-text-input
          v-model="editUser.displayName"
          class="oc-mb-s"
          :label="$gettext('First and last name')"
          :error-message="formData.displayName.errorMessage"
          :fix-message-line="true"
          @input="validateDisplayName"
        />
        <oc-text-input
          v-model="editUser.mail"
          class="oc-mb-s"
          :label="$gettext('Email')"
          :error-message="formData.email.errorMessage"
          type="email"
          :fix-message-line="true"
          @change="validateEmail"
        />
        <oc-text-input
          v-model="editUser.passwordProfile.password"
          class="oc-mb-s"
          :label="$gettext('Password')"
          type="password"
          :fix-message-line="true"
          default-value="●●●●●●●●"
        />
        <div class="oc-mb-s">
          <oc-select
            v-model="editUser.role"
            :label="$gettext('Role')"
            option-label="displayName"
            :options="roles"
            :clearable="false"
          />
          <div class="oc-text-input-message"></div>
        </div>
        <quota-select
          v-if="showQuota"
          :key="'quota-select-' + user.id"
          :title="$gettext('Personal quota')"
          :total-quota="editUser.drive.quota.total || 0"
          :max-quota="spaceQuotas.maxPersonalQuota || 0"
          @selectedOptionChange="changeSelectedQuotaOption"
        />
        <p v-else v-translate class="oc-m-rm">
          To set an individual quota, the user needs to have logged in once.
        </p>
      </div>
    </div>
    <compare-save-dialog
      class="edit-compare-save-dialog"
      :original-object="originalObjectUser"
      :compare-object="compareObjectUser"
      :confirm-button-disabled="invalidFormData"
      @revert="revertChanges"
      @confirm="$emit('confirm', editUser)"
    ></compare-save-dialog>
  </div>
</template>
<script>
import * as EmailValidator from 'email-validator'
import UserInfoBox from './UserInfoBox.vue'
import CompareSaveDialog from 'web-pkg/src/components/sidebar/CompareSaveDialog.vue'
import QuotaSelect from 'web-pkg/src/components/QuotaSelect.vue'
import { cloneDeep } from 'lodash-es'
import { mapState } from 'vuex'

export default {
  name: 'EditPanel',
  components: {
    UserInfoBox,
    CompareSaveDialog,
    QuotaSelect
  },
  props: {
    user: {
      type: Object,
      required: true
    },
    roles: {
      type: Array,
      required: true
    }
  },
  data() {
    return {
      editUser: {},
      formData: {
        displayName: {
          errorMessage: '',
          valid: true
        },
        email: {
          errorMessage: '',
          valid: true
        }
      }
    }
  },
  computed: {
    ...mapState('runtime/spaces', ['spaceQuotas']),

    userRole() {
      return this.user ? this.userRoles[this.user.id] : null
    },

    originalObjectUser() {
      return { ...this.user, passwordProfile: { password: '' } }
    },
    compareObjectUser() {
      return { ...this.editUser }
    },
    invalidFormData() {
      return Object.values(this.formData)
        .map((v) => !!v.valid)
        .includes(false)
    },
    showQuota() {
      return this.editUser.drive
    }
  },
  watch: {
    user: {
      handler: function () {
        this.editUser = { ...cloneDeep(this.user), ...{ passwordProfile: { password: '' } } }
      },
      deep: true,
      immediate: true
    }
  },
  methods: {
    changeSelectedQuotaOption(option) {
      this.editUser.drive.quota.total = option.value
    },

    validateDisplayName() {
      this.formData.displayName.valid = false

      if (this.editUser.displayName.trim() === '') {
        this.formData.displayName.errorMessage = this.$gettext('Display name cannot be empty')
        return false
      }

      this.formData.displayName.errorMessage = ''
      this.formData.displayName.valid = true
      return true
    },

    validateEmail() {
      this.formData.email.valid = false

      if (!EmailValidator.validate(this.editUser.mail)) {
        this.formData.email.errorMessage = this.$gettext('Please enter a valid email')
        return false
      }

      this.formData.email.errorMessage = ''
      this.formData.email.valid = true
      return true
    },

    revertChanges() {
      this.editUser = { ...cloneDeep(this.user), ...{ passwordProfile: { password: '' } } }
      Object.values(this.formData).forEach((formDataValue) => {
        formDataValue.valid = true
        formDataValue.errorMessage = ''
      })
    }
  }
}
</script>
<style lang="scss">
.edit-compare-save-dialog {
  position: absolute;
  bottom: 0;
  left: 0;
}
</style>
