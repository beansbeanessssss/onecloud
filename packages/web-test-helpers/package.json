{
  "name": "web-test-helpers",
  "private": true,
  "main": "src/index.ts",
  "peerDependencies": {
    "axios": "0.27.2",
    "vue-gettext": "2.1.12",
    "vue-router": "3.6.5",
    "vuex": "4.1.0"
  }
}
