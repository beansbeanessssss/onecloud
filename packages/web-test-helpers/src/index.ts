export * from './mocks/defaultComponentMocks'
export * from './mocks/store/defaultStoreMockOptions'
export * from './mocks/defaultStubs'
export * from './mocks/axios'
export * from './defaultPlugins'
export * from './helpers'
