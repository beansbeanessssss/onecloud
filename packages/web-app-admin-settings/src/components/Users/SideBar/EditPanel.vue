<template>
  <div id="user-edit-panel" class="oc-mt-xl">
    <UserInfoBox :user="user" />
    <form id="user-edit-form" class="oc-background-highlight oc-p-m" autocomplete="off">
      <div>
        <oc-text-input
          v-model="editUser.onPremisesSamAccountName"
          class="oc-mb-s"
          :label="$gettext('User name')"
          :error-message="formData.userName.errorMessage"
          :fix-message-line="true"
          @update:model-value="validateUserName"
        />
        <oc-text-input
          v-model="editUser.displayName"
          class="oc-mb-s"
          :label="$gettext('First and last name')"
          :error-message="formData.displayName.errorMessage"
          :fix-message-line="true"
          @update:model-value="validateDisplayName"
        />
        <oc-text-input
          v-model="editUser.mail"
          class="oc-mb-s"
          :label="$gettext('Email')"
          :error-message="formData.email.errorMessage"
          type="email"
          :fix-message-line="true"
          @change="validateEmail"
        />
        <oc-text-input
          :model-value="editUser.passwordProfile?.password"
          class="oc-mb-s"
          :label="$gettext('Password')"
          type="password"
          :fix-message-line="true"
          placeholder="●●●●●●●●"
          @update:model-value="onUpdatePassword"
        />
        <div class="oc-mb-s">
          <oc-select
            :model-value="editUser"
            :label="$gettext('Role')"
            option-label="displayName"
            :options="translatedRoleOptions"
            :clearable="false"
            @update:model-value="onUpdateRole"
          >
            <template #selected-option>
              {{ selectedRoleName }}
            </template>
          </oc-select>
          <div class="oc-text-input-message"></div>
        </div>
        <div class="oc-mb-s">
          <oc-select
            id="login-input"
            :disabled="isLoginInputDisabled"
            :model-value="editUser"
            :label="$gettext('Login')"
            :options="loginOptions"
            :clearable="false"
            @update:model-value="onUpdateLogin"
          >
            <template #selected-option>
              {{ selectedLoginLabel }}
            </template>
          </oc-select>
          <div class="oc-text-input-message"></div>
        </div>
        <quota-select
          v-if="showQuota"
          id="quota-select-form"
          :key="'quota-select-' + user.id"
          class="oc-mb-s"
          :title="$gettext('Personal quota')"
          :total-quota="editUser.drive.quota.total || 0"
          :max-quota="maxQuota"
          @selected-option-change="changeSelectedQuotaOption"
        />
        <p
          v-else
          class="oc-mb-m oc-mt-rm oc-text-meta"
          v-text="$gettext('To set an individual quota, the user needs to have logged in once.')"
        />
        <group-select
          class="oc-mb-s"
          :selected-groups="editUser.memberOf"
          :group-options="groupOptions"
          @selected-option-change="changeSelectedGroupOption"
        />
      </div>
      <compare-save-dialog
        class="edit-compare-save-dialog oc-mb-l"
        :original-object="compareSaveDialogOriginalObject"
        :compare-object="editUser"
        :confirm-button-disabled="invalidFormData"
        @revert="revertChanges"
        @confirm="$emit('confirm', { user, editUser })"
      ></compare-save-dialog>
    </form>
  </div>
</template>
<script lang="ts">
import { computed, defineComponent, PropType, ref, unref } from 'vue'
import * as EmailValidator from 'email-validator'
import UserInfoBox from './UserInfoBox.vue'
import CompareSaveDialog from 'web-pkg/src/components/sideBar/CompareSaveDialog.vue'
import GroupSelect from './GroupSelect.vue'
import QuotaSelect from 'web-pkg/src/components/QuotaSelect.vue'
import { cloneDeep } from 'lodash-es'
import { Group, User } from 'web-client/src/generated'
import { MaybeRef, useGraphClient, useStore } from 'web-pkg'
import { useCapabilitySpacesMaxQuota } from 'web-pkg/src/composables'

export default defineComponent({
  name: 'EditPanel',
  components: {
    UserInfoBox,
    CompareSaveDialog,
    QuotaSelect,
    GroupSelect
  },
  props: {
    user: {
      type: Object as PropType<User>,
      required: false
    },
    roles: {
      type: Array,
      required: true
    },
    groups: {
      type: Array as PropType<Group[]>,
      required: true
    }
  },
  emits: ['confirm'],
  setup(props) {
    const store = useStore()
    const currentUser = store.getters.user
    const editUser: MaybeRef<User> = ref({})
    const formData = ref({
      displayName: {
        errorMessage: '',
        valid: true
      },
      userName: {
        errorMessage: '',
        valid: true
      },
      email: {
        errorMessage: '',
        valid: true
      }
    })
    const groupOptions = computed(() => {
      const { memberOf: selectedGroups } = unref(editUser)
      return props.groups
        .filter((g) => !selectedGroups.some((s) => s.id === g.id))
        .sort((a, b) => a.displayName.localeCompare(b.displayName))
    })

    const isLoginInputDisabled = computed(() => currentUser.uuid === (props.user as User).id)

    return {
      maxQaxQuota: useCapabilitySpacesMaxQuota(),
      isLoginInputDisabled,
      editUser,
      formData,
      groupOptions,
      ...useGraphClient()
    }
  },
  computed: {
    loginOptions() {
      return [
        {
          label: this.$gettext('Allowed'),
          value: true
        },
        {
          label: this.$gettext('Forbidden'),
          value: false
        }
      ]
    },
    selectedLoginLabel() {
      return this.editUser.accountEnabled === false
        ? this.$gettext('Forbidden')
        : this.$gettext('Allowed')
    },
    translatedRoleOptions() {
      return this.roles.map((role) => {
        return { ...role, displayName: this.$gettext(role.displayName) }
      })
    },
    invalidFormData() {
      return Object.values(this.formData)
        .map((v: any) => !!v.valid)
        .includes(false)
    },
    showQuota() {
      return this.editUser.drive?.quota
    },
    compareSaveDialogOriginalObject() {
      return cloneDeep(this.user)
    },
    selectedRoleName() {
      return this.$gettext(
        this.roles.find((role) => role.id === this.editUser.appRoleAssignments[0].appRoleId)
          .displayName
      )
    }
  },
  watch: {
    user: {
      handler: function () {
        this.editUser = cloneDeep(this.user)
      },
      deep: true,
      immediate: true
    },
    editUser: {
      handler: function () {
        /**
         * Property accountEnabled won't be always set, but this still means, that login is allowed.
         * So we actually don't need to change the property if missing and not set to forbidden in the UI.
         * This also avoids the compare save dialog from displaying that there are unsaved changes.
         */
        if (this.editUser.accountEnabled === true && this.user.accountEnabled !== false) {
          delete this.editUser.accountEnabled
        }
      },
      deep: true
    }
  },
  methods: {
    changeSelectedQuotaOption(option) {
      this.editUser.drive.quota.total = option.value
    },
    changeSelectedGroupOption(option: Group) {
      this.editUser.memberOf = option
    },
    async validateUserName() {
      this.formData.userName.valid = false

      if (this.editUser.onPremisesSamAccountName.trim() === '') {
        this.formData.userName.errorMessage = this.$gettext('User name cannot be empty')
        return false
      }

      if (this.editUser.onPremisesSamAccountName.includes(' ')) {
        this.formData.userName.errorMessage = this.$gettext('User name cannot contain white spaces')
        return false
      }

      if (
        this.editUser.onPremisesSamAccountName.length &&
        !isNaN(this.editUser.onPremisesSamAccountName[0])
      ) {
        this.formData.userName.errorMessage = this.$gettext('User name cannot start with a number')
        return false
      }

      if (this.user.onPremisesSamAccountName !== this.editUser.onPremisesSamAccountName) {
        try {
          await this.graphClient.users.getUser(this.editUser.onPremisesSamAccountName)
          this.formData.userName.errorMessage = this.$gettext('User "%{userName}" already exists', {
            userName: this.editUser.onPremisesSamAccountName
          })
          return false
        } catch (e) {
          /**
           * If the backend throws an error, the user doesn't exist and everything is alright
           */
        }
      }

      this.formData.userName.errorMessage = ''
      this.formData.userName.valid = true
      return true
    },
    validateDisplayName() {
      this.formData.displayName.valid = false

      if (this.editUser.displayName.trim() === '') {
        this.formData.displayName.errorMessage = this.$gettext(
          'First and last name cannot be empty'
        )
        return false
      }

      this.formData.displayName.errorMessage = ''
      this.formData.displayName.valid = true
      return true
    },
    validateEmail() {
      this.formData.email.valid = false

      if (!EmailValidator.validate(this.editUser.mail)) {
        this.formData.email.errorMessage = this.$gettext('Please enter a valid email')
        return false
      }

      this.formData.email.errorMessage = ''
      this.formData.email.valid = true
      return true
    },
    revertChanges() {
      this.editUser = cloneDeep(this.user)
      Object.values(this.formData).forEach((formDataValue: any) => {
        formDataValue.valid = true
        formDataValue.errorMessage = ''
      })
    },
    onUpdateRole(role) {
      this.editUser.appRoleAssignments[0].appRoleId = role.id
    },
    onUpdatePassword(password) {
      this.editUser.passwordProfile = {
        password
      }
    },
    onUpdateLogin({ value }) {
      this.editUser.accountEnabled = value
    }
  }
})
</script>

<style lang="scss">
#user-edit-panel {
  #user-edit-form {
    border-radius: 5px;
  }
}
</style>
