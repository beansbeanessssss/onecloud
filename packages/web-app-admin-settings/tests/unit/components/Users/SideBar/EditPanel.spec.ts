import EditPanel from '../../../../../src/components/Users/SideBar/EditPanel.vue'
import { defaultPlugins, shallowMount } from 'web-test-helpers'

describe('EditPanel', () => {
  it('renders all available inputs', () => {
    const { wrapper } = getWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })
  describe('method "revertChanges"', () => {
    it('should revert changes on property editUser', () => {
      const { wrapper } = getWrapper()
      wrapper.vm.editUser.displayName = 'jana'
      wrapper.vm.editUser.mail = 'jana@owncloud.com'
      wrapper.vm.revertChanges()
      expect(wrapper.vm.editUser.displayName).toEqual('jan')
      expect(wrapper.vm.editUser.mail).toEqual('jan@owncloud.com')
    })
    it('should revert changes on property formData', () => {
      const { wrapper } = getWrapper()
      wrapper.vm.formData.displayName.valid = false
      wrapper.vm.formData.displayName.errorMessage = 'error'
      wrapper.vm.revertChanges()
      expect(wrapper.vm.formData.displayName.valid).toBeTruthy()
      expect(wrapper.vm.formData.displayName.errorMessage).toEqual('')
    })
  })

  describe('method "validateDisplayName"', () => {
    it('should return true if displayName is valid', () => {
      const { wrapper } = getWrapper()
      wrapper.vm.editUser.displayName = 'jan'
      expect(wrapper.vm.validateDisplayName()).toBeTruthy()
      expect(wrapper.vm.formData.displayName.valid).toBeTruthy()
    })
    it('should return false if displayName is not valid', () => {
      const { wrapper } = getWrapper()
      wrapper.vm.editUser.displayName = ''
      expect(wrapper.vm.validateDisplayName()).toBeFalsy()
      expect(wrapper.vm.formData.displayName.valid).toBeFalsy()
    })
  })

  describe('method "validateEmail"', () => {
    it('should return true if email is valid', () => {
      const { wrapper } = getWrapper()
      wrapper.vm.editUser.mail = 'jan@owncloud.com'
      expect(wrapper.vm.validateEmail()).toBeTruthy()
      expect(wrapper.vm.formData.email.valid).toBeTruthy()
    })
    it('should return false if email is not valid', () => {
      const { wrapper } = getWrapper()
      wrapper.vm.editUser.mail = ''
      expect(wrapper.vm.validateEmail()).toBeFalsy()
      expect(wrapper.vm.formData.email.valid).toBeFalsy()
    })
  })

  describe('computed method "invalidFormData"', () => {
    it('should be false if formData is invalid', () => {
      const { wrapper } = getWrapper()
      wrapper.vm.formData.displayName.valid = true
      expect(wrapper.vm.invalidFormData).toBeFalsy()
    })
    it('should be true if formData is valid', () => {
      const { wrapper } = getWrapper()
      wrapper.vm.formData.displayName.valid = false
      expect(wrapper.vm.invalidFormData).toBeTruthy()
    })
  })
})

function getWrapper() {
  return {
    wrapper: shallowMount(EditPanel, {
      props: {
        user: {
          id: '1',
          displayName: 'jan',
          mail: 'jan@owncloud.com',
          passwordProfile: { password: '' },
          drive: { quota: {} }
        },
        roles: [{ id: '1', displayName: 'admin' }]
      },
      global: {
        plugins: [...defaultPlugins()]
      }
    })
  }
}
