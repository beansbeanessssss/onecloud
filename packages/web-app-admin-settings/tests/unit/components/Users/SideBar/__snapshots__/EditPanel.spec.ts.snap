// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditPanel renders all available inputs 1`] = `
<div class="oc-mt-xl" id="user-edit-panel">
  <user-info-box-stub user="[object Object]"></user-info-box-stub>
  <form autocomplete="off" class="oc-background-highlight oc-p-m" id="user-edit-form">
    <div>
      <oc-text-input-stub class="oc-mb-s" clearbuttonaccessiblelabel="" clearbuttonenabled="false" disabled="false" errormessage="" fixmessageline="true" id="oc-textinput-1" label="User name" type="text"></oc-text-input-stub>
      <oc-text-input-stub class="oc-mb-s" clearbuttonaccessiblelabel="" clearbuttonenabled="false" disabled="false" errormessage="" fixmessageline="true" id="oc-textinput-2" label="First and last name" modelvalue="jan" type="text"></oc-text-input-stub>
      <oc-text-input-stub class="oc-mb-s" clearbuttonaccessiblelabel="" clearbuttonenabled="false" disabled="false" errormessage="" fixmessageline="true" id="oc-textinput-3" label="Email" modelvalue="jan@owncloud.com" type="email"></oc-text-input-stub>
      <oc-text-input-stub class="oc-mb-s" clearbuttonaccessiblelabel="" clearbuttonenabled="false" disabled="false" fixmessageline="true" id="oc-textinput-4" label="Password" modelvalue="" placeholder="●●●●●●●●" type="password"></oc-text-input-stub>
      <div class="oc-mb-s">
        <oc-select-stub clearable="false" disabled="false" filter="[Function]" id="oc-select-5" label="Role" loading="false" model-value="[object Object]" optionlabel="displayName" options="[object Object]" searchable="true"></oc-select-stub>
        <div class="oc-text-input-message"></div>
      </div>
      <div class="oc-mb-s">
        <oc-select-stub clearable="false" disabled="false" filter="[Function]" id="login-input" label="Login" loading="false" model-value="[object Object]" options="[object Object],[object Object]" searchable="true"></oc-select-stub>
        <div class="oc-text-input-message"></div>
      </div>
      <quota-select-stub class="oc-mb-s" id="quota-select-form" maxquota="0" title="Personal quota" totalquota="0"></quota-select-stub>
      <group-select-stub class="oc-mb-s" groupoptions="undefined,undefined" selectedgroups=""></group-select-stub>
    </div>
    <compare-save-dialog-stub class="edit-compare-save-dialog oc-mb-l" compareobject="[object Object]" confirmbuttondisabled="false" originalobject="[object Object]"></compare-save-dialog-stub>
  </form>
</div>
`;
